constraints.pro

```diff
@@ -276,6 +290,16 @@ gen_enforced_dependency(WorkspaceCwd, DependencyIdent, null, DependencyType) :-
   workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType),
   DependencyType == 'devDependencies'.
 
+% If a controller dependency (other than `base-controller` and
+% `eth-keyring-controller`) is listed under "dependencies", it should also be
+% listed under "peerDependencies". Each controller is a singleton, so we need
+% to ensure the versions used match expectations.
+gen_enforced_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, 'peerDependencies') :-
+  workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, 'dependencies'),
+  DependencyIdent \= '@metamask/base-controller',
+  DependencyIdent \= '@metamask/eth-keyring-controller',
+  is_controller(DependencyIdent).
```

Comment:
Since we're just interested in other workspaces, could we also write this as the following? Might remove the need for `is_controller`.

``` suggestion
gen_enforced_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, 'peerDependencies') :-
  workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, 'dependencies'),
  DependencyIdent \= '@metamask/base-controller',
  workspace_ident(_, DependencyIdent).
```

Reply:
Hmm, I think this would give us the wrong answer in two ways. We don't just care about workspaces here, we care about controllers. Controllers are singletons, regardless whether they live in this repo or not. But the other workspaces aren't. They're libraries that the clients can have multiple different versions of.

Reply:
I was considering keeping it just to controllers in this repo, and adding another constraint to prevent depending on other controllers. Ideally any that are depended on here would be here. I guess we can consider that separately though.